contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,245 @@
+     *         address (Safe address) => address (component address) => EnabledModuleInfo
+     */
+    mapping(address => mapping(address => ModuleAccessInfo)) public enabledModules;
+    struct ModuleAccessInfo {
+        bool rootAddressGranted;
+        address nextModulePointer;
+    }
+
+    // Events
+    event ActionsExecuted(address indexed safe, bytes32 metaHash, uint256 nonce);
+    event RootAccessActionExecuted(address indexed safe, bytes32 metaHash);
+    event ModuleEnabled(address indexed safe, address indexed module, bool allowRootAccess);
+    event ModuleDisabled(address indexed safe, address indexed module);
+
+    // Errors
+    error ModuleRequiresRootAccess(address sender);
+    error MoudleNotEnabled(address module);
+    error ModuleEnabledOnlyForRootAccess(address module);
+    error ModuleAccessMismatch(address module, bool requiresRootAccess, bool providedValue);
+    error ActionExecutionFailed(address safe, bytes32 metaHash, uint256 index);
+    error RootAccessActionExecutionFailed(address safe, bytes32 metaHash);
+    error ModuleAlreadyEnabled(address safe, address module);
+    error InvalidModuleAddress(address module);
+    error InvalidPrevModuleAddress(address module);
+    error ZeroPageSizeNotAllowed();
+
+    modifier onlyEnabledModule(address safe) {
+        if (enabledModules[safe][msg.sender].nextModulePointer == address(0)) {
+            revert MoudleNotEnabled(msg.sender);
+        }
+        _;
+    }
+
+    modifier noZeroOrSentinelModule(address module) {
+        if (module == address(0) || module == SENTINEL_MODULES) {
+            revert InvalidModuleAddress(module);
+        }
+        _;
+    }
+
+    /**
+     * @notice This function executes non-delegate call(s) on a safe if the module is enabled on the Safe.
+     *         If any one of the actions fail, the transaction reverts.
+     * @param safe Address of the Safe instance
+     * @param transaction A struct of type SafeTransaction containing information of about the action(s) to be executed.
+     *                    Users can add logic to validate metahash through a transaction guard.
+     * @return data bytes types containing the result of the executed action.
+     */
+    function executeTransaction(
+        address safe,
+        SafeTransaction calldata transaction
+    ) external override onlyEnabledModule(safe) returns (bytes[] memory data) {
+        data = new bytes[](transaction.actions.length);
+        uint256 length = transaction.actions.length;
+        for (uint256 i = 0; i < length; ++i) {
+            SafeProtocolAction memory safeProtocolAction = transaction.actions[i];
+            (bool isActionSuccessful, bytes memory resultData) = ISafe(safe).execTransactionFromModuleReturnData(
+                safeProtocolAction.to,
+                safeProtocolAction.value,
+                safeProtocolAction.data,
+                0
+            );
```

Comment:
Would it still be possible for the module to execute a transaction with a safe as a recipient? E.g., `execTransactionFromModuleReturnData` -> owner changes, execTransactionFromModule with a delegate call, etc. Don't see a test / require statements for this

Reply:
https://github.com/5afe/safe-protocol-specs/discussions/7